src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1234,6 +1236,86 @@ private IEnumerable<TypeInferenceInfo> InferTypeInPropertyDeclaration(PropertyDe
             return CreateResult(typeInfo.Type);
         }
 
+        private IEnumerable<TypeInferenceInfo> InferTypeInExpressionElement(ExpressionElementSyntax expressionElement)
+        {
+            if (expressionElement.Parent is CollectionExpressionSyntax collectionExpression)
+            {
+                var collectionType = SemanticModel.GetTypeInfo(collectionExpression, CancellationToken).ConvertedType;
+
+                // Try to figure out the type based on the type of hte collection itself.
+                if (CollectionExpressionExtensions.IsConstructibleCollectionType(
+                        SemanticModel.Compilation, collectionType, out var elementTypes) &&
+                    elementTypes is [var elementType, ..])
+                {
+                    return [new(elementType)];
+                }
+
+                // If that fails, see if we can figure out from one of our siblings.
+                foreach (var element in collectionExpression.Elements)
+                {
+                    if (element != expressionElement && element is ExpressionElementSyntax siblingElement)
+                    {
+                        var types = GetTypes(siblingElement.Expression, objectAsDefault: false);
+                        if (types.Any())
+                            return types;
+                    }
+                }
+            }
+
+            return [];
+        }
+
+        private IEnumerable<TypeInferenceInfo> InferTypeInKeyValuePairElement(
+            KeyValuePairElementSyntax keyValuePairElement, ExpressionSyntax expression)
+        {
+            var isKey = expression == keyValuePairElement.KeyExpression;
+            var types = InferTypes();
+            return types.Select(t => new TypeInferenceInfo(t));
+
+            IEnumerable<ITypeSymbol> InferTypes()
+            {
+                if (keyValuePairElement.Parent is CollectionExpressionSyntax collectionExpression)
+                {
+                    var collectionType = SemanticModel.GetTypeInfo(collectionExpression, CancellationToken).ConvertedType;
+
+                    // Try to figure out the type based on the type of hte collection itself.
+                    if (CollectionExpressionExtensions.IsConstructibleCollectionType(
+                            SemanticModel.Compilation, collectionType, out var elementTypes) &&
+                        elementTypes is [var elementType, ..])
```

Comment:
(see other comment for IsConstructibleCollectionType):

I was expecting the fact elementTypes is collection was to potentially allow for this to return two: the key and value here.

Reply:
yeah.  fixed this to just return a single type.